### PR TITLE
Issue #5314: At Time Zone

### LINKED ISSUE #5314: Implement the Postgres timezone function
The [timezone function](https://www.postgresql.org/docs/current/functions-datetime.html#FUNCTIONS-DATETIME-ZONECONVERT) is what the syntactic sugar of `AT TIME ZONE` translates to but we have not implemented it:

```sql
select now() at time zone 'europe/london' as start_time;
Error: Binder Error: No function matches the given name and argument types 'timezone(VARCHAR, TIMESTAMP)'. You might need to add explicit type casts.
    Candidate functions:
    timezone(TIMESTAMP WITH TIME ZONE) -> BIGINT
    timezone(DATE) -> BIGINT
    timezone(TIMESTAMP) -> BIGINT
    timezone(INTERVAL) -> BIGINT
```

What it seems to do is either a) reinterpret a `TIMESTAMP`'s (UTC) bin values as values in the given zone and use them to produce a `TIMESTAMPTZ` or b) produce a TIMESTAMP whose (UTC) bin values are the same as the bin values of the `TIMESTAMPTZ`'s in the given zone.

Basically, it is a mechanism for encoding local time into instants. Which is sketchy but not hard to implement in ICU.

### PR DESCRIPTION
**Pull request by hawkfish**
Implement the underlying TIMEZONE function for this syntax.

fixes #5314

### CODE REVIEW

File: test/sql/timezone/test_icu_timezone.test [hunk 1]
```diff
@@ -18,6 +18,16 @@ TimeZone	America/Los_Angeles	The current time zone	VARCHAR
 statement error
 SET TimeZone = 'fnord';
 
+query I
+SELECT  '2001-02-16 20:38:40'::TIMESTAMP AT TIME ZONE 'America/Denver';
+----
+2001-02-16 19:38:40-08
+
+query I
+SELECT '2001-02-16 20:38:40-05'::TIMESTAMPTZ AT TIME ZONE 'America/Denver';
```
Comment (Mytherin) [inline test/sql/timezone/test_icu_timezone.test:27]: This function seems to be relatively complex but does not have that many tests, could we add tests with:

* NULL as time zone
* NULL as timestamp
* Infinity timestamps
* Non-existent time zone
* All of the above but with timestamps/time zones stored in a table (i.e. not constant)

**Review by Mytherin — COMMENTED** [review]
Thanks for the PR! Looks good. One request for more tests

File: test/sql/timezone/test_icu_timezone.test (hunk 1 above)
Comment (hawkfish) [reply test/sql/timezone/test_icu_timezone.test:27]: Yeah sorry I just copied and pasted a bunch of existing code. Coming shortly.

**Comment by Mytherin** [comment]
Thanks!

**Pull request merged**